File: leetcode/problems/language-specific/go/54.spiral-matrix.go
/*
 * @lc app=leetcode id=54 lang=golang
 *
 * [54] Spiral Matrix
 */

// @lc code=start
func spiralOrder(matrix [][]int) []int {
	if len(matrix) <= 0 || len(matrix[0]) <= 0 {
		return []int{}
	}

	res, m, n := []int{}, len(matrix), len(matrix[0])
	left, right, top, bottom := 0, n-1, 0, m-1
	for left <= right && top <= bottom {
		// * Traverse from top left corner <=> top right corner
		for j := left; j <= right && len(res) < m*n; j++ {
			res = append(res, matrix[top][j])
		}

		// * Traverse from top right corner <=> bottom right corner
		for i := top + 1; i <= bottom && len(res) < m*n; i++ {
			res = append(res, matrix[i][right])
		}

		// * Traverse from bottom right corner <=> bottom left corner
		for j := right - 1; j >= left && len(res) < m*n; j-- {
			res = append(res, matrix[bottom][j])
		}

		// * Traverse from bottom left corner <=> top left corner
		for i := bottom - 1; i > top && len(res) < m*n; i-- {
			res = append(res, matrix[i][left])
		}

		left++
		right--
		top++
		bottom--
	}

	return res
}

// @lc code=end

File: leetcode/problems/problem-specific/54.spiral-matrix/54.spiral-matrix.py
#
# @lc app=leetcode id=54 lang=python3
#
# [54] Spiral Matrix
#

# @lc code=start

# * Iterative Solution | O(mn) Time | O(1) Space
# * m -> The number of rows in the matrix | n -> The number of columns in the matrix


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        left, right = 0, len(matrix[0])
        top, bottom = 0, len(matrix)
        while left < right and top < bottom:
            # * Traverse from the top-left corner to the top-right corner.
            for col in range(left, right):
                res.append(matrix[top][col])
            top += 1

            # * Traverse from the top-right corner to the bottom-right corner.
            for row in range(top, bottom):
                res.append(matrix[row][right - 1])
            right -= 1

            if not left < right or not top < bottom:
                break

            # * Traverse from the bottom-right corner to the bottom-left corner.
            for col in range(right - 1, left - 1, -1):
                res.append(matrix[bottom - 1][col])
            bottom -= 1

            # * Traverse from the bottom-left corner to the top-left corner.
            for row in range(bottom - 1, top - 1, -1):
                res.append(matrix[row][left])
            left += 1

        return res


# @lc code=end
